core/corehttp/hostname.go

```diff
@@ -0,0 +1,301 @@
+									// Redirect to CID fixed inside of toSubdomainURL()
+									http.Redirect(w, r, newURL, http.StatusMovedPermanently)
+									return
+								}
+							}
+						}
+
+						// Rewrite the path to not use subdomains
+						r.URL.Path = pathPrefix + r.URL.Path
+						// Serve path request
+						childMux.ServeHTTP(w, r)
+						return
+					}
+					// If not, resource does not exist on the subdomain gateway, return 404
+					http.NotFound(w, r)
+					return
+				}
+			}
+			// We don't have a known gateway. Fallback on DNSLink lookup
+
+			// Wildcard HTTP Host check:
+			// 1. is wildcard DNSLink enabled (Gateway.NoDNSLink=false)?
+			// 2. does Host header include a fully qualified domain name (FQDN)?
+			// 3. does DNSLink record exist in DNS?
+			if !cfg.Gateway.NoDNSLink && isDNSLinkRequest(r, n) {
+				// rewrite path and handle as DNSLink
+				r.URL.Path = "/ipns/" + stripPort(r.Host) + r.URL.Path
+				childMux.ServeHTTP(w, r)
+				return
+			}
+
+			// else, treat it as an old school gateway, I guess.
+			childMux.ServeHTTP(w, r)
+		})
+		return childMux, nil
+	}
+}
+
+// isDNSLinkRequest returns bool that indicates if request
+// should return data from content path listed in DNSLink record (if exists)
+func isDNSLinkRequest(r *http.Request, n *core.IpfsNode) bool {
+	ctx, cancel := context.WithCancel(n.Context())
+	defer cancel()
+
+	fqdn := stripPort(r.Host)
+	if len(fqdn) > 0 && isd.IsDomain(fqdn) {
+		// Confirm DNSLink was not disabled for the fqdn or globally
+		name := "/ipns/" + fqdn
+		// check if DNSLink exists
+		_, err := n.Namesys.Resolve(ctx, name, nsopts.Depth(1))
+		if err == nil || err == namesys.ErrResolveRecursion {
+			return true
+		}
+	}
+	return false
+}
+
+func isSubdomainNamespace(ns string) bool {
+	switch ns {
+	case "ipfs", "ipns", "p2p", "ipld", "api":
+		return true
+	default:
+		return false
+	}
+}
+
+// Parses Host header of a subdomain-based URL and returns it's components
+// Note: hostname is host + optional port
+func parseSubdomains(hostHeader string) (hostname, ns, rootID string, ok bool) {
+	parts := subdomainGatewayRegex.FindStringSubmatch(hostHeader)
+	if len(parts) < 4 || !isSubdomainNamespace(parts[2]) {
+		return "", "", "", false
+	}
+	return parts[3], parts[2], parts[1], true
+}
+
+// Converts a hostname/path to a subdomain-based URL, if applicable.
+func toSubdomainURL(hostname, path string, query string) (url string, ok bool) {
+	parts := strings.SplitN(path, "/", 4)
+
+	var ns, rootID, rest string
+	switch len(parts) {
+	case 4:
+		rest = parts[3]
+		fallthrough
+	case 3:
+		ns = parts[1]
+		rootID = parts[2]
+	default:
+		return "", false
+	}
+
+	if !isSubdomainNamespace(ns) {
+		return "", false
+	}
+
+	// add prefix if query is present
+	if query != "" {
+		query = "?" + query
+	}
+
+	if ns == "api" || ns == "p2p" {
+		// API and P2P proxy use the same paths on subdomains:
+		// api.hostname/api/.. and p2p.hostname/p2p/..
+		return fmt.Sprintf(
+			"http://%s.%s/%s/%s/%s%s",
```

Comment:
Let's explicitly use the schema from the request. We don't yet support https, but this could turn into a bug in the future.

Reply:
Request does not have schema unless we have TLS listener in go-ipfs itself.
We assume TLS termination at reverse proxy such as Nginx, so in go-ipfs `r.URL.Scheme` is always empty. 

What I will do instead is to follow HTTP best practices for handling scheme in reverse proxy setups: 
- add support for [`X-Forwarded-Proto`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-Proto) header
   -  use `https:` Scheme if `X-Forwarded-Proto: https` header is found in request. 
   - use `http:` Scheme otherwise

Reply:
Done in https://github.com/ipfs/go-ipfs/pull/6096/commits/61d21c280799426dd60c891ad30d560a5ccae5cd